File: comsumer.py
#! -*- encoding: utf-8 -*-
import concurrent.futures
import json
import threading
import time
from queue import Queue

import redis
import requests

r = redis.Redis(host='localhost', port=6380, db=0)

sub = r.pubsub()

topic_handle_map = {}


def handler_test(message):
    # time.sleep(0.01)
    return message


topic_handle_map['test'] = handler_test
topic_handle_map['test1'] = handler_test
topic_handle_map['test2'] = handler_test


def fresh_subscribe():
    while True:
        topics=list(r.smembers('Topics'))
        if len(topics)>0:
            sub.subscribe(topics)
        time.sleep(1)


threading.Thread(target=fresh_subscribe).start()


print(sub.channels)

q = Queue()


def redis_subscribe():
    while True:
        if len(sub.channels)==0:
            time.sleep(0.01)
            continue
        msg = sub.get_message()
        if msg:
            # 去掉"Topic_" 前缀
            topic = msg['channel'][len('Topic_'):]
            q.put(topic)


threading.Thread(target=redis_subscribe).start()


def wrap_result(message, result):
    url = message.get('callback_url')
    if not url:
        return
    message['id'] = message.get('id')
    message['topic'] = message.get('topic')
    message['action'] = 'response'
    # message['message'] =  message.get('id')
    message['message'] = {
        'id': message.get('id'),
        'ts': time.time_ns() - result,
        'origin_ts': result,
    }
    message = json.dumps(message)
    resp = requests.post(url, data=message)
    # print(resp.text)


def consumer():
    while True:
        topic = q.get()
        message = r.rpop(topic)
        if not message:
            continue
        handler = topic_handle_map.get(topic.decode('utf-8', 'ignore'))
        if not handler:
            continue
        origin_message = json.loads(message)
        result = handler(origin_message.get('message'))
        wrap_result(origin_message, result)


# 创建线程池
thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=400)

# 启动多个线程消费队列
for _ in range(40):
    thread_pool.submit(consumer)

# 等待线程池中的所有任务完成
thread_pool.shutdown()

File: client/comsumer.go
package main

import (
	"context"
	"encoding/json"
	"github.com/redis/go-redis/v9"
	"github.com/sjqzhang/requests"
	"log"
	"strings"
	"time"
)

var (
	redisClient *redis.Client
	sub         *redis.PubSub
	topicMap    = make(map[string]func(Message) interface{})
	q           chan string
)

type Message struct {
	Action      string      `json:"action"`
	Topic       string      `json:"topic"`
	Message     interface{} `json:"message"`
	ID          string      `json:"id"`
	CallbackURL string      `json:"callback_url"`
}

func init() {
	redisClient = redis.NewClient(&redis.Options{
		Addr: "127.0.0.1:6380",
		DB:   0,
	})
	sub = redisClient.Subscribe(context.Background(), "Topic_*")
	q = make(chan string, 100)
}

func handlerTest(msg Message) interface{} {
	return msg
}

func handlerProxy(msg Message) interface{} {

	type Req struct {
		Method string          `json:"method"`
		Url    string          `json:"url"`
		Body   string          `json:"body"`
		Header requests.Header `json:"headers"`
	}

	var req Req

	var data []byte

	switch msg.Message.(type) {
	case string:
		data = []byte(msg.Message.(string))
	case []byte:
		data = msg.Message.([]byte)
	default:
		data, _ = json.Marshal(msg.Message)
	}

	err := json.Unmarshal(data, &req)
	if err != nil {
		log.Println(err)
		return err
	}
	resp, err := requests.Requests().Do(strings.ToUpper(req.Method), req.Url, req.Header, req.Body)
	if err != nil {
		return err.Error()
	}
	return resp.Text()

}

func init() {

	topicMap["test"] = handlerTest
	topicMap["test1"] = handlerTest
	topicMap["test2"] = handlerTest
	topicMap["proxy"] = handlerProxy
}

func freshSubscribe() {
	for {
		topics, err := redisClient.SMembers(context.Background(), "Topics").Result()
		if err != nil {
			log.Println(err)
			continue
		}
		if len(topics) > 0 {
			err := sub.Subscribe(context.Background(), topics...)
			if err != nil {
				log.Println(err)
			}
		}
		time.Sleep(1 * time.Second)
	}
}

func redisSubscribe() {
	for {
		message, err := sub.ReceiveMessage(context.Background())
		if err != nil {
			log.Println(err)
			continue
		}
		topic := message.Channel[len("Topic_"):]
		q <- topic
	}
}

func wrapResult(msg Message, result interface{}) {
	msg.Message = result
	msg.Action = "response"
	data, err := json.Marshal(msg)
	if err != nil {
		log.Println(err)
		return
	}
	_, err = requests.PostJson(msg.CallbackURL, string(data))
	if err != nil {
		log.Println(err)
		return
	}
}

func consumer() {
	for {
		topic := <-q
		for {
			message, err := redisClient.RPop(context.Background(), topic).Bytes()
			if err != nil || len(message) == 0 {
				log.Println(err)
				break
			}
			if handler, ok := topicMap[topic]; ok {
				var msg Message
				json.Unmarshal(message, &msg)
				result := handler(msg)
				wrapResult(msg, result)
			}
		}
	}
}

func main() {
	go freshSubscribe()
	time.Sleep(1 * time.Second)
	go redisSubscribe()

	threadPool := make(chan struct{}, 40)
	for i := 0; i < cap(threadPool); i++ {
		threadPool <- struct{}{}
	}

	for {
		select {
		case <-threadPool:
			go func() {
				consumer()
				<-threadPool
			}()
		default:
			consumer()
		}
	}
}
